### ISSUE TITLE
Wrong Number of Arguments When Resolving from Container

### ISSUE THREAD

**Issue by wjdhamilton** [dry-rb/dry-auto_inject#15]
Guys,

This is probably a rookie error, but I am trying to implement `dry-container` with `dry-autoinject` as per http://icelab.com.au/articles/better-code-with-an-inversion-of-control-container/.

Here's my container: 

```ruby
require "dry-container"
require "dry-auto_inject"
require "easy_json_matcher/node"
require "easy_json_matcher/array_generator"
require "easy_json_matcher/attribute_generator"
require "easy_json_matcher/schema_library"


module EasyJSONMatcher
  class Container
    extend Dry::Container::Mixin
  end

  Container.register "node", -> { Node }
  Container.register "array_generator", -> { ArrayGenerator }
  Container.register "attribute_generator", -> { AttributeGenerator }
  Container.register "schema_library", -> { SchemaLibrary }

  AutoInject = Dry::AutoInject(Container)
end
```

and here's the injection: 

```ruby
module EasyJSONMatcher
  class NodeGenerator
    include AutoInject["node", "attribute_generator", "array_generator", "schema_library"]
    include AttributeTypeMethods
```

...which looks right to me (naturally), however when I try to create an instance of the `NodeGenerator` class by calling `#new` I get the following error:

`ArgumentError: wrong number of arguments (given 4, expected 0)`

...don't really know where to start with this. Any input would be appreciated!

**Comment by wjdhamilton** [comment]
This has something to do with the constructor for `NodeGenerator`. If I remove the `#initialize`, it works.

**Comment by flash-gordon** [comment]
That's because dry-auto_inject does something for you behind the scenes. First of all it defines `.new` method for your class where it extracts values from the container and passes them to `#initialize`. The second part is defining `#initialize`, not in your a class but in a mixin so your class now has a constructor with four arguments defined in the mixin. When you want to have your custom constructor along with generated one you just do nothing with arguments and call `super` like that:
```ruby
def initialize(*)
  super
  # my custom logic
end
```

**Comment by wjdhamilton** [comment]
@flash-gordon Thank you very much. Fixed it! There was two things I had to work through: one was that I'm using keyword args and so I had to adapt the call to `AutoInject` to this, as per the tests: 

```ruby
include AutoInject.kwargs[...]
```

and the constructor had to be changed as follows:

```ruby
def initialize(opts: [], global_opts: [], **args)
```

...and then also pull the master branch since the ruby gems version threw a`NoMethodError` for kwargs. 

Now just to investigate how all this magic happens and I'll be able to move along...
